### PR TITLE
Implement Stash API.

### PR DESCRIPTION
**Pull request by Therzok**
Based on https://github.com/libgit2/libgit2/pull/2705

Not in master yet, adding so I can test locally.

### CODE REVIEW

File: LibGit2Sharp/StashCollection.cs [hunk 1]
```diff
@@ -114,6 +114,36 @@ public virtual void Remove(int index)
             Proxy.git_stash_drop(repo.Handle, index);
         }
 
+		/// <summary>
+		/// Apply a single stash to the repository.
+		/// </summary>
+		/// <param name="index">The index of the stash to remove (0 being the most recent one).</param>
+		/// <param name="reinstateIndex">Whether to stage the items or not.</param>
+	    public virtual void Apply(int index, bool reinstateIndex)
```
Comment (dahlbyk) [inline LibGit2Sharp/StashCollection.cs]: :+1: for `StashApplyResult` as we do with merge, cherry-pick, etc.

File: LibGit2Sharp/StashCollection.cs [hunk 2]
```diff
@@ -114,6 +114,36 @@ public virtual void Remove(int index)
...
+
+			Proxy.git_stash_apply(repo.Handle, index, reinstateIndex);
+	    }
+
+		/// <summary>
+		/// Applies a stash to the repository and drops it.
+		/// </summary>
+		/// <param name="index">The index of the stash to remove (0 being the most recent one).</param>
+		/// <param name="reinstateIndex">Whether to stage the items or not.</param>
+		public virtual void Pop(int index, bool reinstateIndex)
```
Comment (dahlbyk) [inline LibGit2Sharp/StashCollection.cs]: See https://github.com/libgit2/libgit2sharp/issues/865#issuecomment-72377143

**Comment by Therzok** [comment]
Okay, I'm advancing on this thing.

I got a few issues I have to point out.

* This won't build on travis until we have that PR merged into upstream.
* https://github.com/libgit2/libgit2/pull/3018#issuecomment-89837236 - I'm testing this specific behavior.
* Conflicts don't appear in the index for some reason.
* Test coverage is horrible. :dancer:

**Comment by Therzok** [comment]
I'm also experiencing some weird behaviour:
 * If I use ReinstateIndex, I get the new changes in the workdir.
 * Something screwed up (I updated GitCheckoutOpts, I probably missed something else?) I'm getting Invalid version 2049 on git_checkout_options).
 * No intermediary updates. :D

File: LibGit2Sharp/StashCollection.cs [hunk 3]
```diff
@@ -114,6 +114,36 @@ public virtual void Remove(int index)
             Proxy.git_stash_drop(repo.Handle, index);
         }
 
+        /// <summary>
+        /// Applies a single stashed state from the stash list
+        /// </summary>
+        /// <param name="index">the index of the stash to remove (0 being the most recent one).</param>
+        /// <param name="options">the options to use for checking out the stash.</param>
+        /// <param name="flags">the flags to use for applying the changes.</param>
+        public virtual StashApplyStatus Apply(int index, StashApplyModifiers flags = StashApplyModifiers.Default, CheckoutOptions options = null)
```
Comment (Therzok) [inline LibGit2Sharp/StashCollection.cs]: Probably this could be abstracted in a StashApplyOptions.

/cc @nulltoken

**Pull request closed without merging**